code/modules/power/reactor/fuel_rods.dm

```diff
@@ -0,0 +1,246 @@
+	if(process)
+		START_PROCESSING(SSobj, src)
+
+/obj/item/fuel_rod/Destroy()
+	if(process)
+		STOP_PROCESSING(SSobj, src)
+	var/obj/machinery/atmospherics/components/trinary/nuclear_reactor/N = loc
+	if(istype(N))
+		N.fuel_rods -= src
+	. = ..()
+
+// This proc will try to convert your fuel rod if you don't override this proc
+// So, ideally, you should write an override of this for every fuel rod you want to create
+/obj/item/fuel_rod/proc/depletion_final(result_rod)
+	if(!result_rod)
+		return
+	var/obj/machinery/atmospherics/components/trinary/nuclear_reactor/N = loc
+	// Rod conversion is moot when you can't find the reactor
+	if(istype(N))
+		var/obj/item/fuel_rod/R
+
+		// You can add your own depletion scheme and not override this proc if you are going to convert a fuel rod into another type
+		switch(result_rod)
+			if("plutonium")
+				R = new /obj/item/fuel_rod/plutonium(loc)
+				R.depletion = depletion
+			if("depleted")
+				if(fuel_power < 10)
+					fuel_power = 0
+					playsound(loc, 'sound/effects/supermatter.ogg', 100, TRUE)
+					R = new /obj/item/fuel_rod/depleted(loc)
+					R.depletion = depletion
+
+		// Finalization of conversion
+		if(istype(R))
+			N.fuel_rods += R
+			qdel(src)
+	else
+		depleted_final = FALSE // Maybe try again later?
+
+/obj/item/fuel_rod/proc/deplete(amount=0.035)
+	depletion += amount * depletion_speed_modifier
+	if(depletion >= depletion_threshold && !depleted_final)
+		depleted_final = TRUE
+		depletion_final(depletion_conversion_type)
+
+/obj/item/fuel_rod/plutonium
+	fuel_power = 0.20
+	name = "plutonium-239 fuel rod"
+	desc = "A highly energetic titanium sheathed rod containing a sizeable measure of weapons grade plutonium, it's highly efficient as nuclear fuel, but will cause the reaction to get out of control if not properly utilised."
+	icon_state = "inferior"
+	rad_strength = 1500
+	process = TRUE // for half life code
+	depletion_threshold = 300
+	depletion_conversion_type = "depleted"
+
+/obj/item/fuel_rod/process()
+	fuel_power = og_fuel_power * 0.5**((world.time - time_created) / half_life SECONDS) // halves the fuel power every half life (33 minutes)
+
+/obj/item/fuel_rod/depleted
+	fuel_power = 0.05
+	name = "depleted fuel rod"
+	desc = "A highly radioactive fuel rod which has expended most of its useful energy."
+	icon_state = "normal"
+	rad_strength = 6000 // smelly
+	depletion_conversion_type = null // It means that it won't turn into anything
+	process = TRUE
+
+// Master type for material optional (or requiring, wyci) and/or producing rods
+/obj/item/fuel_rod/material
+	// Whether the rod has been harvested. Should be set in expend().
+	var/expended = FALSE
+	// The material that will be inserted and then multiplied (or not). Should be some sort of /obj/item/stack
+	var/material_type
+	// The name of material that'll be used for texts
+	var/material_name
+	var/material_name_singular
```

Comment:
`/obj/item/stack` should have `singular_name` defined

TC may not have them but you can add them